File: qrs-core-js/src/transfer.ts
/**
 * Transfer envelope.
 *
 * QR codes are the offline medium for moving signed objects (TCert, SDoc,
 * Statement) between devices — desktop to mobile, or vice-versa. The payload is a
 * small, scannable, self-describing string so the receiving app knows what it is
 * and what to do with it before parsing:
 *
 *     qrs://v1/<type>/<base64url-bytes>
 *
 * This is a *transport* format only: the receiving app still performs the full
 * protocol steps (parse, verify signature, resolve trust, check revocation).
 *
 * Multiple objects can travel together in a *bundle* (e.g. a CA shares an
 * attestation *with* the complete attested TCert so the verifier can fully
 * process it offline):
 *
 *     qrs://v1/bundle/<base64url-json>
 */
import { fromBase64Url, toBase64Url } from './id.js';

export type TransferObjectType = 'tcert' | 'sdoc' | 'statement';

export const TRANSFER_SCHEME = 'qrs';
export const TRANSFER_VERSION = 'v1';

export interface DecodedTransferPayload {
  type: TransferObjectType;
  bytesB64: string;
}

/** Build a transfer payload for a signed object's base64url bytes. */
export function encodeTransferPayload(type: TransferObjectType, bytesB64: string): string {
  return `${TRANSFER_SCHEME}://${TRANSFER_VERSION}/${type}/${bytesB64}`;
}

/** Parse a transfer payload. Returns `null` when the payload is not one of ours. */
export function decodeTransferPayload(payload: string): DecodedTransferPayload | null {
  const prefix = `${TRANSFER_SCHEME}://${TRANSFER_VERSION}/`;
  if (!payload.startsWith(prefix)) return null;
  const rest = payload.slice(prefix.length);
  const slash = rest.indexOf('/');
  if (slash === -1) return null;
  const type = rest.slice(0, slash) as TransferObjectType;
  const bytesB64 = rest.slice(slash + 1);
  if (type !== 'tcert' && type !== 'sdoc' && type !== 'statement') return null;
  if (bytesB64.length === 0) return null;
  return { type, bytesB64 };
}

/* -------------------------------------------------------------------------- */
/* Bundle (multiple signed objects in one payload)                             */
/* -------------------------------------------------------------------------- */

export interface BundleObject {
  type: TransferObjectType;
  bytesB64: string;
}

export interface DecodedBundle {
  objects: BundleObject[];
}

/** Encode several signed objects as one scannable payload (e.g. TCert + attestation). */
export function encodeBundle(objects: BundleObject[]): string {
  const json = JSON.stringify({ v: TRANSFER_VERSION, objects });
  const b64 = toBase64Url(new TextEncoder().encode(json));
  return `${TRANSFER_SCHEME}://${TRANSFER_VERSION}/bundle/${b64}`;
}

/** Parse a bundle payload. Returns `null` when the payload is not a bundle. */
export function decodeBundle(payload: string): DecodedBundle | null {
  const prefix = `${TRANSFER_SCHEME}://${TRANSFER_VERSION}/bundle/`;
  if (!payload.startsWith(prefix)) return null;
  const b64 = payload.slice(prefix.length);
  try {
    const json = JSON.parse(new TextDecoder().decode(fromBase64Url(b64))) as {
      v?: string;
      objects?: unknown;
    };
    if (!Array.isArray(json.objects)) return null;
    const objects: BundleObject[] = [];
    for (const o of json.objects as Array<Record<string, unknown>>) {
      if (!o || typeof o.type !== 'string' || typeof o.bytesB64 !== 'string') continue;
      const type = o.type as TransferObjectType;
      if (type !== 'tcert' && type !== 'sdoc' && type !== 'statement') continue;
      objects.push({ type, bytesB64: o.bytesB64 });
    }
    return { objects };
  } catch {
    return null;
  }
}

/* -------------------------------------------------------------------------- */
/* `.qrs` file container                                                        */
/* -------------------------------------------------------------------------- */

/**
 * File extension used for QRS export files (a plain-text container whose content
 * is exactly a transfer payload — `qrs://v1/<type>/<base64url>` or a bundle).
 * This lets a user save a TCert / SDoc / Statement to a `.qrs` file, share it
 * out-of-band (WhatsApp, email, USB), and have the receiver import it by feeding
 * the file text straight through `processPayload`.
 */
export const QRS_FILE_EXTENSION = 'qrs';

/** Build the UTF-8 bytes of a `.qrs` file holding a single signed object. */
export function encodeQrsFile(type: TransferObjectType, bytesB64: string): Uint8Array {
  return new TextEncoder().encode(encodeTransferPayload(type, bytesB64));
}

/** Build the UTF-8 bytes of a `.qrs` file holding a bundle of signed objects. */
export function encodeQrsBundleFile(objects: BundleObject[]): Uint8Array {
  return new TextEncoder().encode(encodeBundle(objects));
}

export type DecodedQrsFile =
  | { kind: 'object'; payload: DecodedTransferPayload }
  | { kind: 'bundle'; objects: BundleObject[] };

/** Parse the text content of a `.qrs` file back into an object or a bundle. */
export function decodeQrsFile(text: string): DecodedQrsFile | null {
  const trimmed = text.trim();
  if (!trimmed) return null;
  const bundle = decodeBundle(trimmed);
  if (bundle) return { kind: 'bundle', objects: bundle.objects };
  const payload = decodeTransferPayload(trimmed);
  if (payload) return { kind: 'object', payload };
  return null;
}

File: qrs-core-python/src/qrs/transfer.py
"""Transfer envelope.

QR codes are the offline medium for moving signed objects (TCert, SDoc,
Statement) between devices — desktop to mobile, or vice-versa. The payload is a
small, scannable, self-describing string so the receiving app knows what it is
and what to do with it before parsing:

    qrs://v1/<type>/<base64url-bytes>

This is a *transport* format only: the receiving app still performs the full
protocol steps (parse, verify signature, resolve trust, check revocation).

Multiple objects can travel together in a *bundle* (e.g. a CA shares an
attestation *with* the complete attested TCert so the verifier can fully
process it offline):

    qrs://v1/bundle/<base64url-json>
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from .constants import QRS_FILE_EXTENSION, TRANSFER_SCHEME, TRANSFER_VERSION
from .id import from_base64url, to_base64url

__all__ = [
    "TRANSFER_SCHEME",
    "TRANSFER_VERSION",
    "QRS_FILE_EXTENSION",
    "TRANSFER_OBJECT_TYPES",
    "is_transfer_object_type",
    "encode_transfer_payload",
    "decode_transfer_payload",
    "encode_bundle",
    "decode_bundle",
    "encode_qrs_file",
    "encode_qrs_bundle_file",
    "decode_qrs_file",
]

TRANSFER_OBJECT_TYPES: tuple[str, ...] = ("tcert", "sdoc", "statement")


def is_transfer_object_type(value: Any) -> bool:
    return value in TRANSFER_OBJECT_TYPES


@dataclass
class DecodedTransferPayload:
    type: str
    bytes_b64: str


@dataclass
class BundleObject:
    type: str
    bytes_b64: str


@dataclass
class DecodedBundle:
    objects: list[BundleObject]


def encode_transfer_payload(obj_type: str, bytes_b64: str) -> str:
    """Build a transfer payload for a signed object's base64url bytes."""
    return f"{TRANSFER_SCHEME}://{TRANSFER_VERSION}/{obj_type}/{bytes_b64}"


def decode_transfer_payload(payload: str) -> DecodedTransferPayload | None:
    """Parse a transfer payload. Returns ``None`` when it is not one of ours."""
    prefix = f"{TRANSFER_SCHEME}://{TRANSFER_VERSION}/"
    if not payload.startswith(prefix):
        return None
    rest = payload[len(prefix):]
    slash = rest.find("/")
    if slash == -1:
        return None
    obj_type = rest[:slash]
    bytes_b64 = rest[slash + 1:]
    if not is_transfer_object_type(obj_type):
        return None
    if len(bytes_b64) == 0:
        return None
    return DecodedTransferPayload(type=obj_type, bytes_b64=bytes_b64)


def encode_bundle(objects: list[BundleObject]) -> str:
    """Encode several signed objects as one scannable payload (e.g. TCert + attestation).

    Accepts either :class:`BundleObject` instances or plain dicts with
    ``type``/``bytes_b64`` (or ``bytesB64``) keys, mirroring the reference
    implementation's ergonomics.
    """
    entries = []
    for o in objects:
        if isinstance(o, BundleObject):
            entries.append({"type": o.type, "bytesB64": o.bytes_b64})
        elif isinstance(o, dict):
            obj_type = o.get("type")
            bytes_b64 = o.get("bytes_b64", o.get("bytesB64"))
            entries.append({"type": obj_type, "bytesB64": bytes_b64})
        else:
            raise TypeError(f"Unsupported bundle object: {type(o).__name__}")
    json_text = json.dumps(
        {"v": TRANSFER_VERSION, "objects": entries},
        separators=(",", ":"),
    )
    b64 = to_base64url(json_text.encode("utf-8"))
    return f"{TRANSFER_SCHEME}://{TRANSFER_VERSION}/bundle/{b64}"


def decode_bundle(payload: str) -> DecodedBundle | None:
    """Parse a bundle payload. Returns ``None`` when the payload is not a bundle."""
    prefix = f"{TRANSFER_SCHEME}://{TRANSFER_VERSION}/bundle/"
    if not payload.startswith(prefix):
        return None
    b64 = payload[len(prefix):]
    try:
        decoded = json.loads(from_base64url(b64).decode("utf-8"))
    except Exception:  # noqa: BLE001 - malformed bundle is simply "not a bundle"
        return None
    if not isinstance(decoded, dict) or not isinstance(decoded.get("objects"), list):
        return None
    objects: list[BundleObject] = []
    for o in decoded["objects"]:
        if not isinstance(o, dict):
            continue
        obj_type = o.get("type")
        bytes_b64 = o.get("bytesB64")
        if not isinstance(obj_type, str) or not isinstance(bytes_b64, str):
            continue
        if not is_transfer_object_type(obj_type):
            continue
        objects.append(BundleObject(type=obj_type, bytes_b64=bytes_b64))
    return DecodedBundle(objects=objects)


def encode_qrs_file(obj_type: str, bytes_b64: str) -> bytes:
    """Build the UTF-8 bytes of a ``.qrs`` file holding a single signed object."""
    return encode_transfer_payload(obj_type, bytes_b64).encode("utf-8")


def encode_qrs_bundle_file(objects: list[BundleObject]) -> bytes:
    """Build the UTF-8 bytes of a ``.qrs`` file holding a bundle of signed objects."""
    return encode_bundle(objects).encode("utf-8")


# DecodedQrsFile is a tagged union: ``kind`` is "object" or "bundle".
DecodedQrsFile = dict[str, Any]


def decode_qrs_file(text: str) -> DecodedQrsFile | None:
    """Parse the text content of a ``.qrs`` file back into an object or a bundle."""
    trimmed = text.strip()
    if not trimmed:
        return None
    bundle = decode_bundle(trimmed)
    if bundle is not None:
        return {"kind": "bundle", "objects": bundle.objects}
    payload = decode_transfer_payload(trimmed)
    if payload is not None:
        return {"kind": "object", "payload": payload}
    return None